core-addon/IonCore.js

```diff
@@ -103,6 +124,45 @@ module.exports = class IonCore {
     }
   }
 
+  /**
+   * Handles messages coming in from studies.
+   *
+   * IMPORTANT: This is `async` and will always handle all the
+   * external messages coming from the studies. It will not allow
+   * any other registered listener to catch messages. See
+   * the notes at the listener registration site.
+   *
+   * @param {Object} message
+   *        The payload of the message.
+   * @param {runtime.MessageSender} sender
+   *        An object containing informations about who sent
+   *        the message.
+   * @returns {Promise} The response to the received message.
+   *          It can be resolved with a value that is sent to the
+   *          `sender`.
+   */
+  async _handleExternalMessage(message, sender) {
+    // We only expect messages coming from known ion studies.
+    let knownStudies = await this._availableStudies;
+    if (!knownStudies.map(s => s.addon_id).includes(sender.id)) {
+      return Promise.reject(
+        new Error(`IonCore._handleExternalMessage - unexpected sender ${sender.id}`));
+    }
+
+    switch (message.type) {
+      case "telemetry-ping": {
+        const data = message.data;
+        return await this._dataCollection.sendPing(
+          data.payloadType, data.payload, data.namespace,
+          data.keyId, data.key
```

Comment:
This is a nit feel free to ignore :) 

```suggestion
        const {payloadType, payload, namespace, keyId, key} = message.data;
        return await this._dataCollection.sendPing(
          payloadType, payload, namespace,
          keyId, key
```

Reply:
You could also just destructure inside the function call but I think that is probably harder to read:
```
        return await this._dataCollection.sendPing(
          {payloadType, payload, namespace,
          keyId, key} = message.data
```

Reply:
Oh! That looks great! Let me do what you recommended in the first comment!